File: Frontend/lib/services/userService.ts
import { Role, User, UserAuth } from "../types/userService";
import ApiClient from "./apiClient";

const port = 10010;
const client = new ApiClient(port);

const UserService = {
  login: client.api<{ userId: string; password: string }, UserAuth>(
    "LoginMessage"
  ),

  logout: client.api<{ token: string }, void>("LogoutMessage"),

  checkAuth: client.api<{ token: string; role: Role }, unknown>(
    "CheckAuthMessage"
  ),

  checkUserRole: client.api<{ userId: string; role: Role }, void>(
    "CheckUserRoleMessage"
  ),

  createUser: client.api<
    {
      adminToken: string;
      userId: string;
      password: string;
      name: string;
      role: Role;
    },
    void
  >("CreateUserMessage"),

  queryUsers: client.api<{ role: Role }, User[]>("QueryUsersMessage"),

  updateUserName: client.api<
    {
      adminToken: string;
      userId: string;
      name: string;
    },
    void
  >("UpdateUserNameMessage"),

  updateUserPassword: client.api<
    {
      adminToken: string;
      userId: string;
      password: string;
    },
    void
  >("UpdateUserPasswordMessage"),

  deleteUser: client.api<
    {
      adminToken: string;
      userId: string;
    },
    void
  >("DeleteUserMessage"),
};

export default UserService;

File: Backend/Core/src/main/scala/APIs/UserService.scala
package APIs

import Common.API.API
import Global.UserService
import Objects._

import io.circe.generic.auto._

case class LoginMessage(userId: String, password: String)
    extends API[UserAuth](UserService, "LoginMessage")

case class LogoutMessage(token: String)
    extends API[Unit](UserService, "LogoutMessage")

case class CheckAuthMessage(token: String, role: Role)
    extends API[String](UserService, "CheckAuthMessage")

case class CheckUserRoleMessage(
    userId: String,
    role: Role
) extends API[Unit](UserService, "CheckUserRoleMessage")

case class CreateUserMessage(
    adminToken: String,
    userId: String,
    password: String,
    name: String,
    role: Role
) extends API[Unit](UserService, "CreateUserMessage")

case class QueryUsersMessage(role: Role)
    extends API[List[User]](UserService, "QueryUsersMessage")

case class UpdateUserPasswordMessage(
    adminToken: String,
    userId: String,
    password: String
) extends API[Unit](UserService, "UpdateUserPasswordMessage")

case class UpdateUserNameMessage(
    adminToken: String,
    userId: String,
    name: String
) extends API[Unit](UserService, "UpdateUserNameMessage")

case class DeleteUserMessage(adminToken: String, userId: String)
    extends API[Unit](UserService, "DeleteUserMessage")
